lib/elixir/lib/calendar/datetime.ex

```diff
@@ -998,6 +998,68 @@ defmodule DateTime do
     naive_diff + System.convert_time_unit(offset_diff, :second, unit)
   end
 
+  @doc """
+  Adds a specified amount of time to a `DateTime`.
+
+  Accepts an `integer` in any `unit` available from `t:System.time_unit/0`.
+  Negative values will move backwards in time.
+
+  Takes changes such as summer time/DST into account. This means that adding time
+  can cause the wall time to "go backwards" during "fall back" during autumn.
+  Adding just a few seconds to a datetime just before "spring forward" can cause wall
+  time to increase by more than an hour.
+
+  Fractional second precision stays the same in a similar way to `NaiveDateTime.add/2`.
+
+  ### Examples
+
+      iex> dt = DateTime.from_naive!(~N[2018-11-15 10:00:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> {:ok, dt_added} = dt |> DateTime.add(3600, :second, FakeTimeZoneDatabase)
+      iex> dt_added
+      #DateTime<2018-11-15 11:00:00+01:00 CET Europe/Copenhagen>
+
+      iex> dt = DateTime.from_naive!(~N[2018-11-15 10:00:00], "Etc/UTC")
+      iex> {:ok, dt_added} = dt |> DateTime.add(3600, :second)
+      iex> dt_added
+      #DateTime<2018-11-15 11:00:00Z>
+
+  When adding 3 seconds just before "spring forward" we go from 1:59:59 to 3:00:02
+
+      iex> dt = DateTime.from_naive!(~N[2019-03-31 01:59:59.123], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> {:ok, dt_added} = dt |> DateTime.add(3, :second, FakeTimeZoneDatabase)
+      iex> dt_added
+      #DateTime<2019-03-31 03:00:02.123+02:00 CEST Europe/Copenhagen>
+
+  """
+  @doc since: "1.8.0"
+  @spec add(t, integer, System.time_unit(), Calendar.get_time_zone_database()) ::
+          {:ok, t()} | {:error, atom}
```

Comment:
I have noticed an API consistency issue. The other `add` functions do not return tuples, but the value, and this would be annoyingly inconsistent. We have two options:

1. Name this function something else
2. Raise in case of `{:error, _}`

It seems the only possible error is due to using the UTC-only database, so I would vote to go with 2.

Reply:
Yes, I made it use tuples to be consistent with other functions in DateTime where the time zone database returns an error if it does not recognize the time zone or cannot handle it (in the case of the UTC only one).

Raising instead of tuples make sense to me. In some of the other function where a time zone name is provided separately it is more likely that it is an invalid time zone name. If the time zone name is from the field in the `DateTime` struct it is probably from the same time zone database the developer is using so it is less likely to be invalid.